tests/common/plugins/pdu_controller/pdu_manager.py

```diff
@@ -0,0 +1,264 @@
+"""
+    PduManager is intended to solve the issue where DUT connects to
+    multiple PDU controllers.
+
+    It also intended to hide the dependency on the fake outlet_id,
+    and reference outlet buy outlet dictionary directly. With this,
+    we could enable different way to identify outlet, e.g. with the
+    outlet number from graph.
+
+    It also intended to create a smooth transition from defining
+    PDU in inventory to defining PDU in connection graph. Data in
+    graph is preferred, but if graph data is missing, existing
+    inventory data will be used.
+
+    PDU manager implements the same base PDU controller APIs and
+    collect status from and distribute operations to individual PDU
+    controllers.
+"""
+
+import logging
+import copy
+from snmp_pdu_controllers import get_pdu_controller
+
+logger = logging.getLogger(__name__)
+
+class PduManager():
+    def __init__(self, dut_hostname):
+        """
+            dut_hostname is the target DUT host name. The dut
+            defines which PDU(s) and outlet(s) it connected to.
+
+            It is NOT the PDU host name. PDU host name is defined
+            either in graph or in inventory and associated with
+            the DUT.
+        """
+        self.dut_hostname = dut_hostname
+        """
+            controlers is an array of controller dictionaries with
+            following information:
+            {
+                'psu_name'   : name of the PSU on DUT,
+                'host'       : controller_IP_address,
+                'controller' : controller instance,
+                'outlets'    : cached outlet status,
+                'psu_peer'   : psu peer information,
+            }
+        """
+        self.controllers = []
+
+
+    def _update_outlets(self, outlets, pdu_index):
+        for outlet in outlets:
+            outlet['pdu_index'] = pdu_index
+            outlet['pdu_name'] = self.controllers[pdu_index]['psu_peer']['peerdevice']
+
+
+    def add_controller(self, psu_name, psu_peer, pdu_vars):
+        """
+            Add a controller to be managed.
+            Sampel psu_peer:
+            {
+                "peerdevice": "pdu-107",
+                "HwSku": "Sentry",
+                "Protocol": "snmp",
+                "ManagementIp": "10.0.0.107",
+                "Type": "Pdu",
+                "peerport": "39"
+            }
+        """
+        if 'Protocol' not in psu_peer or 'ManagementIp' not in psu_peer:
+            logger.info('psu_peer {} missing critical inforamtion'.format(psu_peer))
+            return
+
+        if psu_peer['Protocol'] != 'snmp':
+            logger.warning('Controller protocol {} is not supported'.format(protocol))
+            return
+
+        controller = None
+        pdu_ip = psu_peer['ManagementIp']
+        dup_ctrl = False
```

Comment:
What is `dup_ctrl` short for? Typo?

Reply:
The variable name is not clear. Changed to shared_pdu (basically means that the adding outlet shares the same pdu controller with previous outlet).